invoices/invoiceregistry.go

```diff
@@ -939,12 +939,18 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 
 	// Execute locked notify exit hop logic.
 	i.Lock()
-	resolution, err := i.notifyExitHopHtlcLocked(&ctx, hodlChan)
+	resolution, invoiceToExpire, err := i.notifyExitHopHtlcLocked(
+		&ctx, hodlChan,
+	)
 	i.Unlock()
 	if err != nil {
 		return nil, err
 	}
 
+	if invoiceToExpire != nil {
+		i.expiryWatcher.AddInvoices(invoiceToExpire)
```

Comment:
With this outside the lock, isn't it possible that something happens like the cancellation of an invoice right before the invoice is added to the `expiryWatcher`? Or any of the other invoice events such as a settle?

Reply:
I guess that was already possible previously because the expiry watcher has its own goroutine too. Probably that watcher cancel will be a no-op in that case?

Reply:
Yes it's possible but the watcher handles such cases (https://github.com/lightningnetwork/lnd/blob/95c270d1f88059f2052c3b21ce6df08dd9499f9c/invoices/invoice_expiry_watcher.go#L348)